packages/admin-panel-server/src/__tests__/viz-builder/dashboardVisualisation/DashboardVisualisationExtractor.test.ts

```diff
@@ -0,0 +1,417 @@
+/*
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ *
+ */
+
+import { yup } from '@tupaia/utils';
+
+import {
+  draftDashboardItemValidator,
+  legacyDashboardItemValidator,
+  draftReportValidator,
+  legacyReportValidator,
+} from '../../../viz-builder/dashboardVisualisation/validators';
+import { PreviewMode } from '../../../viz-builder/types';
+import { DashboardVisualisationExtractor } from '../../../viz-builder/dashboardVisualisation/DashboardVisualisationExtractor';
+
+describe('DashboardVisualisationExtractor', () => {
+  describe('validation', () => {
+    it('throws error if viz does not have data field', () => {
+      expect(
+        () => new DashboardVisualisationExtractor({ presentation: {} }, yup.object(), yup.object()),
+      ).toThrow('data is a required field');
+    });
+
+    it('throws error if viz does not have presentation field', () => {
+      expect(
+        () => new DashboardVisualisationExtractor({ data: {} }, yup.object(), yup.object()),
+      ).toThrow('presentation is a required field');
+    });
+
+    describe('getDraftReport', () => {
+      it('throws error if viz does not have data.fetch', () => {
+        expect(() =>
+          new DashboardVisualisationExtractor(
+            { code: 'viz', data: { transform: [] }, presentation: {} },
+            yup.object(),
+            draftReportValidator,
+          ).getReport(),
+        ).toThrow('fetch is a required field');
```

Comment:
In the past I'd follow this format too and try to combine as many things as possible in the `expect` statement. I changed my approach after I read https://docs.microsoft.com/en-us/dotnet/core/testing/unit-testing-best-practices#arranging-your-tests, and I think I prefer it that way now. I had a discussion with @edmofro that Tamanu devs also follow an explicit Arrange/Act/Assert structure in their tests.

Nothing to change here, of course 🙂 Our codebase is filled with this assertion style, just food for thought!

Reply:
Oh wow, I'd never heard of this but it's actually such a good suggestion! Thanks for raising it 😄 I'm happy to change these, and will try to remember this rule in the future 👍